File: simpliMvvm/src/main/java/com/trayis/simplimvvm/ui/SimpliFragment.java
package com.trayis.simplimvvm.ui;

import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.trayis.simplimvvm.utils.Logging;
import com.trayis.simplimvvm.utils.SimpliProviderUtil;
import com.trayis.simplimvvm.viewmodel.SimpliViewModel;

import java.util.InvalidPropertiesFormatException;

public abstract class SimpliFragment<B extends ViewDataBinding> extends Fragment implements SimpliBase {

    protected final String TAG = getClass().getSimpleName();

    private SimpliViewModel[] mViewModel;
    private B mBinding;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mViewModel = getViewModel();
        setHasOptionsMenu(false);
    }

    @SuppressWarnings("unchecked")
    private SimpliViewModel[] getViewModel() {
        if (mViewModel == null) {
            try {
                mViewModel = SimpliProviderUtil.getInstance().getProvider().getViewModels(this);
            } catch (InvalidPropertiesFormatException e) {
                Logging.e(TAG, e.getMessage(), e);
            }
        }
        return mViewModel;
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        mBinding = DataBindingUtil.inflate(inflater, getLayoutResourceId(), container, false);
        return mBinding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mBinding.setLifecycleOwner(this);
        if (mViewModel != null) {
            for (SimpliViewModel viewModel : mViewModel) {
                viewModel.onCreate();
            }
        }
        mBinding.executePendingBindings();
    }

}

File: simpliMvvm/src/main/java/com/trayis/simplimvvm/ui/SimpliFragment.kt
package com.trayis.simplimvvm.ui

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.fragment.app.Fragment
import com.trayis.simplimvvm.ui.SimpliBase.Companion.SELF
import com.trayis.simplimvvm.utils.SimpliProviderUtil
import com.trayis.simplimvvm.viewmodel.SimpliViewModel

abstract class SimpliFragment<B : ViewDataBinding>() : Fragment(), SimpliBase {

    protected val TAG = javaClass.simpleName

    protected var mViewModels: Array<SimpliViewModel?>? = null

    protected var mBinding: B? = null

    private var flag = SELF

    override fun setViewModelProviderFlag(flag: Int) {
        this.flag = flag
    }

    override fun getViewModelProviderFlag() = flag

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mViewModels = SimpliProviderUtil.getProvider()?.getViewModels(this)
        setHasOptionsMenu(false)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        val binding: B = DataBindingUtil.inflate(inflater, layoutResourceId, container, false)
        mBinding = binding
        return binding.getRoot()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        bindViewModels()
        mBinding?.setLifecycleOwner(this)
        mViewModels?.forEach { it?.onCreate() }
        mBinding?.executePendingBindings()
    }

}